`storage`: decompress batch during `batch_timequery()`

WillemKauf opened the pull request:
Fixes https://github.com/redpanda-data/redpanda/issues/25769.

A `ListOffsets` timequery can currently return incorrect results due to the fact that we would not iterate over the records of a compressed batch when constructing the response (this dates back to https://github.com/redpanda-data/redpanda/pull/6658).

We would instead output the same offset for each record in the batch hit by the timequery according to the base offset of the batch itself.


To fix this behavior, decompress the batch and iterate over the records just like we would in the uncompressed case to Do The Right Thing and return correct results.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.1.x
- [X] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Bug Fixes

* Fixes a bug in which the broker response to a `ListOffsets` timequery would return incorrect record offsets when considering a compressed record batch.

WillemKauf committed:
`storage`: make `disk_log_impl::timequery()` a coroutine

This was a particular eyesore to look at in the continuation form.

Make it a co-routine.

WillemKauf committed:
`cloud_storage`: remove intermediate `std::variant` usage

No functional changes.

`model::consume_reader_to_memory()` only returns a
`record_batch_reader::data_t`, so there's no need to store it in
an intermediate `storage_t` variant just to call `std::get<data_t>`
on it immediately afterwards.

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/3236d2de-8c51-4230-8b56-bb3bc2116c6c?refs=292fd176199320d4f5c0a9a356a270aa3826b43f&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/3236d2de-8c51-4230-8b56-bb3bc2116c6c/license?refs=292fd176199320d4f5c0a9a356a270aa3826b43f&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/88c74c9f-2f93-4c05-8855-abe1ffcdc4de?refs=292fd176199320d4f5c0a9a356a270aa3826b43f&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/88c74c9f-2f93-4c05-8855-abe1ffcdc4de/license?refs=292fd176199320d4f5c0a9a356a270aa3826b43f&source=prChecksComment)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65081-->
<!--3fc60f05-404b-44e1-a9d4-e17d0b61651b-->
<!--title-->
## Retry command for [Build#65081](https://buildkite.com/redpanda/redpanda/builds/65081)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":false,"with_chunked_compaction":false,"with_iceberg":false,"with_tiered_storage":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0833be8f-be6d-4fed-8f9d-d22b21b1301b-->
<!--TEST RESULTS-->
<!--18ae81b4-8ff8-4e10-96ad-82cd4b26714c-->
<!--TEST RESULTS-->
<!--965d227e-00a3-4446-8446-03c436322e4e-->
<!--TEST RESULTS-->
<!--e153be23-da7e-44b0-bf11-346224ce3139-->
<!--TEST RESULTS-->
<!--d0aa3189-4c76-49cb-aa9a-2b7d47fcfcdd-->
<!--TEST RESULTS-->
<!--b659742a-6659-45b8-a5fc-a2fb993594fd-->
<!--TEST RESULTS-->
<!--ad06baa1-4e13-46e9-be51-d81b3919882e-->
<!--TEST RESULTS-->
<!--4f2aeb4d-93cc-46a5-9ecc-1d936c496031-->
<!--TEST RESULTS-->
<!--84af54e3-7507-4d05-9cf9-8ec9afc60aaf-->
<!--TEST RESULTS-->
<!--6fa56ae5-60e6-417e-af02-3ebda2735e2d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65081">build#65081</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/65081#01966b60-e81d-43bc-99c0-a4af3a94b000 | FLAKY | 1/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65081#01966b60-e81c-4403-bcba-386de6378fef | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65081#01966b60-e81d-43bc-99c0-a4af3a94b000 | FAIL | 0/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65081#01966ba4-c0d9-40df-b002-c311323471bd | FLAKY | 20/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65081#01966bb8-f74e-46f7-8815-803e033b810d | FLAKY | 17/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65112">build#65112</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_timeboxed_uploads.acks=-1.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65112#01966d7c-e12d-4c82-9fac-029ca0735243 | FLAKY | 20/21 |
| rptest.tests.availability_test.AvailabilityTests.test_recovery_after_catastrophic_failure | ducktape | https://buildkite.com/redpanda/redpanda/builds/65112#01966d7c-e12f-4691-995e-172842025020 | FLAKY | 20/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65112#01966d7c-e131-4d12-a0ac-2ba277291640 | FLAKY | 8/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/65112#01966d69-ab2a-4e39-91e7-e0a5c684d7ac | FLAKY | 13/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65187">build#65187</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_all_moves_in_cluster | ducktape | https://buildkite.com/redpanda/redpanda/builds/65187#01967cfa-149e-4929-8d41-2dbae404ab1a | FLAKY | 20/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/65187#01967d30-321a-4175-a830-ff721fa665bb | FLAKY | 13/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65187#01967d30-3219-4831-b456-5201371e4533 | FLAKY | 7/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65187#01967cfa-14a1-4e43-a816-9b013d2c340a | FLAKY | 12/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65206">build#65206</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65206#01967f6f-8620-442a-bdce-708768f99d6a | FLAKY | 9/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65240">build#65240</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65240#01968280-5486-49de-9abf-ff7aac1a785f | FLAKY | 9/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65298">build#65298</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65298#01968702-5c4b-44ce-8c43-4076144063a5 | FLAKY | 6/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65315">build#65315</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65315#01968797-e2e3-496f-8123-c8603b882669 | FLAKY | 7/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65317">build#65317</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65317#019687b2-972a-414b-aef0-c99a27a103e1 | FLAKY | 7/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65332">build#65332</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65332#01968843-9384-4a83-a002-611fc6a5d46e | FLAKY | 6/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65345">build#65345</a></summary>


| test_id                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit        | https://buildkite.com/redpanda/redpanda/builds/65345#019688c9-08c0-4152-a86d-993c5de4c028 | FLAKY         | 1/2      |
</details>
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65112-->
<!--15325f39-5eed-4e49-951d-d1c4626976ef-->
<!--f0d9cf9e-9a26-410e-9313-1c3521af2b0e-->
<!--d417e74d-b3ce-436f-8595-674474c0eded-->
<!--b5dc8025-7935-4103-aade-86bba259bf7f-->
<!--title-->
## Retry command for [Build#65112](https://buildkite.com/redpanda/redpanda/builds/65112)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/archival_test.py::ArchivalTest.test_timeboxed_uploads@{"acks":-1,"cloud_storage_type":1}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/availability_test.py::AvailabilityTests.test_recovery_after_catastrophic_failure
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
```
<!--/body-->

rockwotj commented on src/v/storage/log_reader.cc:
```diff
@@ -595,22 +596,22 @@ timequery_result batch_timequery(
-              }
-          });
-    }
+    auto batch = internal::decompress_batch_sync(std::move(b));
```
Shouldn't we be making this async?

WillemKauf replied:
I originally wrote it like this to keep `batch_timequery()` a synchronous function, but since it's only called from async functions to begin with, why not.

WillemKauf committed:
`storage`: decompress batch during `batch_timequery()`

`ListOffsets` can currently return incorrect results due to the fact
we would not iterate over the records of a compressed batch. We would
instead output the same offset for each record in the batch hit by the
timequery according to the base offset of the batch itself.

To fix this behavior, decompress the batch and iterate over the records
just like we would in the uncompressed case to do the right thing and
return correct results.

This also has the side-effect of making `batch_timequery()` an asynchronous
function.

WillemKauf commented:
Force push to:
* Make `batch_timequery()` an asynchronous function 
* Re-organize commit history

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65187-->
<!--97ae54af-0d84-48ee-b406-49a7d89608a4-->
<!--5c18b28a-c9de-46cb-96dd-399a5de44f4b-->
<!--2ef3eff7-7f53-4ea3-ad3a-72dd336ed197-->
<!--14ab47c7-5534-405d-9696-8ff48953320e-->
<!--title-->
## Retry command for [Build#65187](https://buildkite.com/redpanda/redpanda/builds/65187)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/partition_move_interruption_test.py::PartitionMoveInterruption.test_cancelling_all_moves_in_cluster
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
tests/rptest/tests/partition_move_interruption_test.py::PartitionMoveInterruption.test_cancelling_all_moves_in_cluster
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65201-->
<!--9b5a0dab-f39c-4af0-a7e9-37168c842a71-->
<!--title-->
## Retry command for [Build#65201](https://buildkite.com/redpanda/redpanda/builds/65201)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

rockwotj commented on src/v/storage/log_reader.cc:
```diff
@@ -595,23 +596,25 @@ timequery_result batch_timequery(
+              return ss::make_ready_future<ss::stop_iteration>(
+                ss::stop_iteration::yes);
+          } else {
+              return ss::make_ready_future<ss::stop_iteration>(
```
Doesn't need to change, but you can also do something like `ss::as_ready_future(ss::stop_iteration::no);`

nvartolomei replied:
ssx::now(ss::stop_iteration::no)? :)

WillemKauf commented:
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65206-->
<!--255f0c67-4da3-4fb9-83f2-d1fb22ada84c-->
<!--title-->
## Retry command for [Build#65206](https://buildkite.com/redpanda/redpanda/builds/65206)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65240-->
<!--79d1eeba-4f06-401f-99d3-1e698e65bb26-->
<!--title-->
## Retry command for [Build#65240](https://buildkite.com/redpanda/redpanda/builds/65240)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

nvartolomei reviewed (approved):
I do wonder why it wasn't implemented earlier. We don't call batch_timequery in tight loops so the overhead of decompressing a batch here and there doesn't seem critical.

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65298-->
<!--a229c9d5-b920-4073-9007-1539b06eab0f-->
<!--title-->
## Retry command for [Build#65298](https://buildkite.com/redpanda/redpanda/builds/65298)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

rockwotj commented:
> I do wonder why it wasn't implemented earlier. We don't call batch_timequery in tight loops so the overhead of decompressing a batch here and there doesn't seem critical.

I don't know why John didn't do it "right" the first time. Maybe @dotnwat remembers? We generally don't decompress outside compaction, so maybe he was being cautious... Some workloads certainly are timequery heavy, and usually it's OK to get a lower bound here, but yeah we've seen issues with timequeries pop up recently (see: https://github.com/redpanda-data/redpanda/pull/25200).

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65315-->
<!--4466c182-4294-4751-9f2e-b621bdf29b6b-->
<!--title-->
## Retry command for [Build#65315](https://buildkite.com/redpanda/redpanda/builds/65315)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65317-->
<!--45191e72-4efc-4b28-abcc-ad076af4d5a8-->
<!--title-->
## Retry command for [Build#65317](https://buildkite.com/redpanda/redpanda/builds/65317)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65332-->
<!--4bc96c57-dabf-4f4f-9b20-697f57f8fa21-->
<!--title-->
## Retry command for [Build#65332](https://buildkite.com/redpanda/redpanda/builds/65332)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25932-v24.3.x-793 remotes/upstream/v24.3.x
git cherry-pick -x 293b17cc36 ac352f4c40 292fd17619
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14767497266).
